### ISSUE TITLE
Typescript breakpoints no longer work since 1.25

### ISSUE THREAD

**Issue by gombosg** [Microsoft/vscode#54026]
- VSCode Version: 1.25 & 1.26 Insiders. 1.24.1 is working properly.
- OS Version: Fedora 28 & Mac OS High Sierra (same symptoms)

Steps to Reproduce:

1. Start our Node.js server app via `launch.json` or attach to a process started from the terminal.
2. Try to set a breakpoint in the server `.ts` files.

Does this issue occur when all extensions are disabled?:
Yes, it's occurring with a vanilla Insiders Edition, too.

Our Typescript directory structure is the following:

- `src/...` - Typescript source files
- `build/...` - Built Javascript files
- `build/src/...` - Source maps

E.g. `src/app.ts` is the main source file, `build/app.js` is the Javascript, `build/src/app.js.map` is the source map.
The Javascript files point to the source maps at their end with a relative path e.g. 
```javascript
//# sourceMappingURL=../src/server/app.js.map`.
```
The source maps are built with gulp and they point to the exact absolute source path at their end e.g. 
```json
"sourceRoot": "/home/gombosg/myApp/src"`
```
`outFiles` is properly set in `launch.json`, but simply attaching to a running process worked, too.

This used to work perfectly, breakpoints always worked, but they appear as 'unverified' (greyed and hollow) since 1.25.  Downgrading back to 1.24.1 helped though.

Let me know if you need some special logs.

**Comment by gombosg** [comment]
Well, it may not be connected to VSCode itself (but e.g. a Node debug adapter component) but something that has been working so far did break on a certain VSCode version update.

**Comment by gombosg** [comment]
I made a simple test case which works in 1.24 but doesn't work in 1.25.
[tsTest.zip](https://github.com/Microsoft/vscode/files/2184182/tsTest.zip)

Instructions:
- Run `build.sh`
- Set a breakpoint in `build/hello.ts`
- Run with `F5`, it's already configured.
- For me it stopped at the breakpoint in 1.24, didn't stop in 1.25.

**Comment by weinand** [comment]
@gombosg does your test case work in the latest Insiders?

**Comment by gombosg** [comment]
@weinand Nope. 1.25 and 1.26 Insiders seem to be affected.

**Comment by roblourens** [comment]
Apparently with the mapRoot and sourceRoot tsconfig options, tsc writes out absolute local paths for sourceMappingURL and sourceRoot. I think this is wrong, they should be file URIs and are usually relative paths. I just "fixed" this last month so that / urls can be resolved from the root of a webserver. Since tsc is doing this, I'll have to support this format too.

**Comment by roblourens** [comment]
In my defense, Chrome devtools doesn't handle windows-style local paths which is what TS outputs on Windows, so I will file a bug on Typescript. It does on MacOS though.

**Comment by roblourens** [comment]
I filed Microsoft/TypeScript#25587

**Comment by gombosg** [comment]
@roblourens Thanks a lot! Also, tsc can be configured from gulp etc. and absolute paths can be set anytime. I actually set absolutes source map paths in the gulp config file because it didn't work other way, it was trial and error. ☺️

**Comment by gombosg** [comment]
FYI as a quick fix I set the source map URLs in Gulp to relative by setting a relative `sourceRoot` in the appropriate [`sourcemaps.write`](https://github.com/gulp-sourcemaps/gulp-sourcemaps#write-options) command. This works with 1.25 properly. (But unlike the example I attached before, it's not a native `tsconfig.json` solution.)

**Comment by paustint** [comment]
I am also having the same issue. I was working on two typescript projects with different code-bases that have been working with breakpoints for at least the past year and both of them stopped working with the update.  :(

I am not using gulp and don't plan to implement it for this issue.  Any idea when this might be fixed so I can plan accordingly?

**Comment by paustint** [comment]
It appears that if I remove `mapRoot` from my `tsconfig.json` the breakpoints are properly handled and I can now debug again! :)

Here are my configs in case others experience issues to aid in possible troubleshooting.

## Before
```
{
  "extends": "../tsconfig.json",
  "compilerOptions": {
    "declaration": false,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "lib": [
      "es7",
      "dom"
    ],
    "mapRoot": "../dist/server",
    "module": "commonjs",
    "moduleResolution": "node",
    "outDir": "../dist/server",
    "sourceMap": true,
    "strictNullChecks": false,
    "noUnusedLocals": false,
    "target": "es6",
    "typeRoots": [
      "../node_modules/@types",
      "./types"
    ]
  }
}
```
## After
```
{
  "extends": "../tsconfig.json",
  "compilerOptions": {
    "declaration": false,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "lib": [
      "es7",
      "dom"
    ],
    "module": "commonjs",
    "moduleResolution": "node",
    "outDir": "../dist/server",
    "sourceMap": true,
    "strictNullChecks": false,
    "noUnusedLocals": false,
    "target": "es6",
    "typeRoots": [
      "../node_modules/@types",
      "./types"
    ]
  }
}
```

**Comment by roblourens** [comment]
This will be fixed in TS' end in TS 3.0, but I'm trying to get a fix out on our end in Insiders.

**Comment by paustint** [comment]
I think the issue I was experiencing was actually from issue #53785 and might be unrelated from what was reported in this ticket.

**Comment by roblourens** [comment]
@paustint Since you said it's related to mapRoot, this is probably the right issue.

**Comment by roblourens** [comment]
Please try this workaround for mac/linux. Add this to your launch config:

```json
"pathMapping": {
    "/": "/"
}
```

You will get an warning that the property is "not allowed", ignore it.

**Comment by noambonnie** [comment]
@gombosg Life saver! changed path to relative in gulp and I have my breakpoints again. Thank you!

**Comment by jamietre** [comment]
Also having this problem - adding `pathMapping` has no effect for me (having this problem on mac). I'm not using gulp, though. Latest insiders doesn't work either. I'm running Jest tests using `ts-jest`.

**Comment by jamietre** [comment]
Never mind my comment above. This is actually a Jest problem for me. Unfortunately both vscode and Jest seem to have problems related to debugging in recent release. For me, the latest release of vscode works with jest@22 + ts-jest but not jest@23. 

https://github.com/facebook/jest/issues/6683

**Comment by roblourens** [comment]
There will be a fix in tomorrow's insiders build, please try it out.

**Comment by JacksonKearl** [comment]
I wasn't able to break with:

tsconfig.json:
```
{
    "include": [
        "src/**/*"
    ],
    "compilerOptions": {
        "sourceMap": true,
        "sourceRoot": "src/",
        "outDir": "build/",
        "mapRoot": "build/"
    }
}
```

but was after removing the mapRoot and sourceRoot entries.

**Comment by JacksonKearl** [comment]
3.0.1

**Comment by roblourens** [comment]
Yeah with those options you get `//# sourceMappingURL=../src/build/foo.js.map` which is nonsense

**Comment by JacksonKearl** [comment]
Verified for TS@<3.0
